### PR TITLE
Lesson7 1

### PR DESCRIPTION
**Pull request by mhope21**
This is lesson 7

### CODE REVIEW

**Comment by raquii** [comment]
I am not sure how, but your model, migrations, and controllers are missing from this PR, and I am realizing now that they were also missing from lesson 6, but I _saw them_ when we met for our 1:1. Are you not including those in your commits? 🤔

File: lesson7-questions.txt [hunk 1]
```diff
@@ -66,15 +79,26 @@ new_subscription = user.subscriptions.new
 Try that last line.  You will see that it creates a subscription object, but that
 object is not entirely empty.  Which attribute is set? Where does it get that
 value?
+When you create a new subcription instance, Rails sets the user_id attribute of the new
+subcription object to the id of the user instance, this is because the subscription model
+belongs to a user and Rails associates the new subcription with that specific user.
```
Comment (dsinn) [inline lesson7-questions.txt]: This answers the second part of the question. The first part is:

> What kind of object is returned by user.subscriptions?

Could you provide the answer?

File: lesson7-questions.txt [hunk 2]
```diff
@@ -66,15 +79,26 @@ new_subscription = user.subscriptions.new
...
 
 
 7. Suppose that user is an instance of User, and forum is an instance of Forum.  What
 Active Record Statement could you use to retrieve all the user's posts for that
 forum?  Hint: You need to have a where in your statement. Try this out in the console.
 You will see the SQL that is executed logged to the screen.  What is that SQL?
 
-
+  SELECT "forums".* FROM "forums" INNER JOIN "subscriptions" ON "forums"."id" = "subscriptions"."forum_id" 
+  WHERE "subscriptions"."user_id" = ?  [["user_id", 1]]
```
Comment (dsinn) [inline lesson7-questions.txt]: The first part of question asks for the Active Record statement, but this is only SQL. The SQL also returns forum records joined with subscriptions, while the question is asking for post records. Could you revise the SQL statement and provide the Active Record statement?

File: lesson7-questions.txt [hunk 3]
```diff
@@ -87,19 +111,37 @@ You will see the SQL that is executed logged to the screen.  What is that SQL?
 There is a security exposure, a risk that a hacker could do damage.  What is that
 risk? What two things are done in your code to prevent the risk?
 
+The risk is that someone could delete a post that was created by someone else. That should not
+be allowed. The code should check if the person is logged on and if it is their post before they
+are allowed to delete it.
+I think this is the relevant code:
+before_action :set_post, only: %w[show edit update destroy]
+  before_action :check_access, only: %w[edit update delete]
+Any idea why one says destroy and one says delete, even though I have no method called delete?
```
Comment (dsinn) [inline lesson7-questions.txt:126]: This was a mistake in the starter code. In idiomatic Rails, `destroy` is the correct method to define.

File: lesson7-questions.txt [hunk 4]
```diff
@@ -87,19 +111,37 @@ You will see the SQL that is executed logged to the screen.  What is that SQL?
...
+It also sets the post on destroy to make sure you are deleting the right one, correct?
```
Comment (dsinn) [inline lesson7-questions.txt:127]: Correct.

File: lesson7-questions.txt [hunk 5]
```diff
@@ -87,19 +111,37 @@ You will see the SQL that is executed logged to the screen.  What is that SQL?
...
 
 
 9. The routes for subscriptions are nested under the forum routes.  How does this
 change the routes?  What is the additional hash key and value that is passed in the
 params hash to the subscriptions controller as a result of this nesting? Which routes
 are unaffected by the nesting?
 
+When you nest subscriptions under forums, the routes become hierarchical, meaning the routes for
+subcriptions will include the forum context in their URL structure. In this case, the new and create route
+will include the forum_id, but the show, edit, update,and destroy will not because of the shallow: true?
```
Comment (dsinn) [inline lesson7-questions.txt:137]: Yes, the `shallow` option is "smart" about which CRUD operations require a forum_id and which ones don't (there's an example on [their docs](https://guides.rubyonrails.org/routing.html#shallow-nesting)).

File: lesson7-questions.txt [hunk 6]
```diff
@@ -87,19 +111,37 @@ You will see the SQL that is executed logged to the screen.  What is that SQL?
...
+The index route is unaffected by the nesting because of except: [:index]? This is the relevant code, I believe:
```
Comment (dsinn) [inline lesson7-questions.txt:138]: Correct, the `except` excludes the actions in the array from being created under the nesting.

**Review by dsinn — CHANGES_REQUESTED** [review]
All of the code looks good, so well done! However, I'd just like to see revised answers for questions 6 and 7 (see other comments).

File: lesson7-questions.txt (hunk 1 above)
Comment (mhope21) [reply lesson7-questions.txt]: Is the answer not that a new subscription object is created? Or are you looking for something different?

Comment (mhope21) [reply lesson7-questions.txt]: I see my mistake now. I was only focused on the second part of the question where it asked about user.subscriptions.new. I will answer the first part of the question.

File: lesson7-questions.txt [hunk 7]
```diff
@@ -75,6 +100,11 @@ forum?  Hint: You need to have a where in your statement. Try this out in the co
 You will see the SQL that is executed logged to the screen.  What is that SQL?
 
 
+user_posts_in_forum = Post.where(user_id: user.id, forum_id: forum.id)
```
Comment (dsinn) [inline lesson7-questions.txt:103]: This works. An alternative solution could make use of the association methods, like so:

```ruby
user.posts.where(forum: forum)
```

**Review by dsinn — APPROVED** [review]

**Pull request merged**